code/modules/stock_market/stocks.dm

```diff
@@ -0,0 +1,307 @@
+		share_change += outside_shareholders * avg_change
+
+	var/cv = last_value
+	supplyDrop(share_change)
+	available_shares += share_change // temporary
+
+	if (prob(25))
+		average_optimism = max(min(average_optimism + (rand(-3, 3) - current_trend * 0.15) / 100, 1), -1)
+
+	var/aspec = abs(speculation)
+	if (prob((aspec - 75) * 2))
+		speculation += rand(-4, 4)
+	else
+		if (prob(50))
+			speculation += rand(-4, 4)
+		else
+			speculation += rand(-400, 0) / 1000 * speculation
+			if (prob(1) && prob(5)) // pop that bubble
+				speculation += rand(-4000, 0) / 1000 * speculation
+
+	current_value += (speculation / rand(25000, 50000) + performance / rand(100, 800)) * current_value
+	if (current_value < 5)
+		current_value = 5
+
+	if (performance != 0)
+		performance = rand(900,1050) / 1000 * performance
+		if (abs(performance) < 0.2)
+			performance = 0
+
+	disp_value_change = (cv < current_value) ? 1 : ((cv > current_value) ? -1 : 0)
+	last_value = current_value
+	if (values.len >= 50)
+		values.Cut(1,2)
+	values += current_value
+
+	if (current_value < 10)
+		unifyShares()
+
+	last_trend = current_trend
+	current_trend += rand(-200, 200) / 100 + optimism * rand(200) / 10 + max(50 - abs(speculation), 0) / 50 * rand(0, 200) / 1000 * (-current_trend) + max(speculation - 50, 0) * rand(0, 200) / 1000 * speculation / 400
+
+/datum/stock/proc/unifyShares()
+	for (var/I in shareholders)
+		var/shr = shareholders[I]
+		if (shr % 2)
+			sellShares(I, 1)
+		shr -= 1
+		shareholders[I] /= 2
+		if (!shareholders[I])
+			shareholders -= I
+	for (var/datum/borrow/B in borrow_brokers)
+		B.share_amount = round(B.share_amount / 2)
+		B.share_debt = round(B.share_debt / 2)
+	for (var/datum/borrow/B in borrows)
+		B.share_amount = round(B.share_amount / 2)
+		B.share_debt = round(B.share_debt / 2)
+	average_shares /= 2
+	available_shares /= 2
+	current_value *= 2
+	last_unification = ticker.round_elapsed_ticks
+
+/datum/stock/process()
+	for (var/datum/borrow/borrow in borrows)
+		if (ticker.round_elapsed_ticks > borrow.grace_expires)
+			modifyAccount(borrow.borrower, -max(current_value * borrow.share_debt, 0), 1)
+			borrows -= borrow
+			if (borrow.borrower in FrozenAccounts)
+				FrozenAccounts[borrow.borrower] -= borrow
+				if (length(FrozenAccounts[borrow.borrower]) == 0)
+					FrozenAccounts -= borrow.borrower
+			qdel(borrow)
+		else if (ticker.round_elapsed_ticks > borrow.lease_expires)
+			if (borrow.borrower in shareholders)
+				var/amt = shareholders[borrow.borrower]
+				if (amt > borrow.share_debt)
+					shareholders[borrow.borrower] -= borrow.share_debt
+					borrows -= borrow
+					if (borrow.borrower in FrozenAccounts)
+						FrozenAccounts[borrow.borrower] -= borrow
+					if (length(FrozenAccounts[borrow.borrower]) == 0)
+						FrozenAccounts -= borrow.borrower
+					qdel(borrow)
+				else
+					shareholders -= borrow.borrower
+					borrow.share_debt -= amt
+	if (bankrupt)
+		return
+	for (var/datum/borrow/borrow in borrow_brokers)
```

Comment:
should only be borrow datums in borrow_brokers.

I'm going to leave this bit here, there's loads of opportunities for istypeless for loops here, fix them.